### DISCUSSION TITLE
BaseHTTPMiddleware + unhandled exceptions cause ResourceWarning about unclosed anyio.streams.memory.MemoryObjectReceiveStream

### DISCUSSION THREAD

**Discussion by nikitagashkov** [encode/starlette#2772] [Potential Issue]
Hello and thank you for your wonderful framework!

I'd like to start a discussion about something I've stumbled upon during testing a `starlette`-based application. We have a FastAPI application with a custom unhandled exception handler that we test in our `pytest`-based suite

After upgrading `anyio` to 4.4.0+ we started recieving failed suites due to `ResourceWarning` being raised from `anyio` (this PR introduced the warning — https://github.com/agronholm/anyio/pull/715 — that've landed in 4.4.0). It seems like that there's nothing wrong in that `anyio` patchset in particular and it just highlighted the issue within `starlette`

I've sprinked `print`s and `tracepack.print_stack()`s within our code to find the origin of the unclosed stream and traces lead to the `BaseHTTPMiddleware`, specifically this line, allocating the `recv_stream`: https://github.com/encode/starlette/blob/0.41.3/starlette/middleware/base.py#L112

A quick research lead me to [this block of code](https://github.com/encode/starlette/blob/0.41.3/starlette/middleware/base.py#L156-164) and it looks like it does not cleanup the `recv_stream` upon unhandled exceptions.

I've come up with the script to reproduce the issue:

```python
# /// script
# requires-python = "==3.13.*"
# dependencies = [
#     "anyio==4.6.2.post1",
#     "certifi==2024.8.30",
#     "h11==0.14.0",
#     "httpcore==1.0.7",
#     "httpx==0.28.0",
#     "idna==3.10",
#     "sniffio==1.3.1",
#     "starlette==0.41.3",
# ]
# ///
import asyncio
import warnings
import tracemalloc
from typing import Sequence, Never

import httpx

from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import PlainTextResponse, Response
from starlette.routing import Route

tracemalloc.start()
warnings.simplefilter("error")


async def ok(_: Request) -> PlainTextResponse:
    return PlainTextResponse("OK")


async def error(_: Request) -> Never:
    raise RuntimeError("Oops!")


class NoopMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        return await call_next(request)


def create_app(middleware: Sequence[Middleware]) -> Starlette:
    return Starlette(
        routes=[
            Route("/ok", endpoint=ok),
            Route("/error", endpoint=error),
        ],
        middleware=middleware,
    )


def create_client(app: Starlette) -> httpx.AsyncClient:
    return httpx.AsyncClient(
        transport=httpx.ASGITransport(app, raise_app_exceptions=False),
        base_url="https://test/",
    )

async def run_test():
    # 1. If no unhandled exceptions are raised, the receiving stream is closed
    app = create_app(middleware=[Middleware(NoopMiddleware)])
    async with create_client(app) as client:
        resp = await client.get("/ok")
        assert resp.status_code == 200
        assert resp.text == "OK"
    del app  # force cleanup to release resources

    # 2. If no middleware is provided, the receiving stream is closed
    app = create_app(middleware=[])
    async with create_client(app) as client:
        resp = await client.get("/error")
        assert resp.status_code == 500
        assert resp.text == "Internal Server Error"
    del app  # force cleanup to release resources

    # 3. If middleware is provided and an unhandled exception is raised, the receiving stream is left open
    app = create_app(middleware=[Middleware(NoopMiddleware)])
    async with create_client(app) as client:
        resp = await client.get("/error")
        assert resp.status_code == 500
        assert resp.text == "Internal Server Error"
    del app  # force cleanup to release resources

if __name__ == '__main__':
    asyncio.run(run_test())
```

Running the following script raises the following issue:

```shell
% uv run anyio-starlette-resource-warning.py
Reading inline script metadata from `anyio-starlette-resource-warning.py`
Installed 8 packages in 15ms
Exception ignored in: <function MemoryObjectReceiveStream.__del__ at 0x1033fb7e0>
Traceback (most recent call last):
  File "/Users/gashkov/.cache/uv/archive-v0/ejHFCA2OM9aoPIAEGr5es/lib/python3.13/site-packages/anyio/streams/memory.py", line 183, in __del__
    warnings.warn(
ResourceWarning: Unclosed <MemoryObjectReceiveStream at 10354d6d0>
```

Probably, this issue might cause warnings for some folks from this discussion as well: https://github.com/encode/starlette/discussions/2603

Thank you for reviewing this report!

**Comment by Kludex** [comment]
Would you try to open a PR to fix it?

**Reply by nikitagashkov** [reply]
Yep, I'll try

**Reply by nikitagashkov** [reply]
Hey @Kludex,

Opened a PR for this — https://github.com/encode/starlette/pull/2777 — please take a look. While I was fixing the issue, found another case that was not so straightforward to fix and changes are more invasive — https://github.com/encode/starlette/pull/2777/commits/8385b641d8f9c64b45d8eeaaf8e0c9d237d8923e — not sure, if it's the best way to tackle this, please let me know if you would like some adjustments 🤔

Thanks!

**Reply by Kludex** [reply]
Can you separate into two PRs then?

**Reply by nikitagashkov** [reply]
Sure, moved this commit from #2777 to #2778, I'll rebase it if #2777 will be accepted
